### ISSUE TITLE
Need more infos about /ZW and why it can't be distributed

### ISSUE THREAD

**Issue by poppolopoppo** [fastbuild/fastbuild#438]
Found in [ObjectNode.cpp/ObjectNode::DetermineFlags](https://github.com/fastbuild/fastbuild/blob/e366ce67ae7a9a36fc6e90cf8ab06fa5e416bd10/Code/Tools/FBuild/FBuildCore/Graph/ObjectNode.cpp#L863) :
````cpp
        // 1) clr code cannot be distributed due to a compiler bug where the preprocessed using
        // statements are truncated
        // 2) code consuming the windows runtime cannot be distributed due to preprocessing weirdness
        // 3) pch files can't be built from preprocessed output (disabled acceleration), so can't be distributed
        // 4) user only wants preprocessor step executed
        if ( !usingCLR && !usingPreprocessorOnly )
        {
            if ( isDistributableCompiler && !usingWinRT && !( flags & ObjectNode::FLAG_CREATING_PCH ) )
            {
                flags |= ObjectNode::FLAG_CAN_BE_DISTRIBUTED;
            }

            // TODO:A Support caching of 7i format
            if ( ( flags & ObjectNode::FLAG_USING_PDB ) == 0 )
            {
                flags |= ObjectNode::FLAG_CAN_BE_CACHED;
            }
        }
`````

Could please give more infos about the weirdness you were talking about ?
Any plan on supported this ? 

Could this limitation be dodged by overriding compiler family to 'custom' and enabling simple distribution ?

This is bothering since it's preventing UWP/XB1 builds from being distributed, since they both need WinRT :(

Maybe I could help if you have an idea of what is missing to get this to work 😉 

Thanks

**Comment by poppolopoppo** [comment]
Sadly just using simple distribution with 'custom' compiler family didn't work 😢

**Comment by ffulin** [comment]
FASTBuild uses the preprocessor to flatten the file include hierarchy into a single file which can be sent to the remote machine for compilation. When using /ZW, the pre-processed output is generated in a way that it isn't actually valid and fails to compile. I don't remember the precise errors I encountered, but you could find out by locally disabling this check. I didn't look at this problem for a long time, so it's probably worth another look - it would definitely be great to support.

For what it's worth, when I originally encountered this problem, it was on an XBox One title and at that time we discovered that /ZW was also bloating all out object files and generally degrading compiler performance, so we isolated any use of the Windows runtime (of which there were not many) into individual files which we compiled with /ZW (while compiling everything else without /ZW). It might not be possible to do this with your code base, and it also might not be a problem with more recent versions of the compiler, but this might be worth a look as an alternative.

With regard to "simple distribution" - that won't work here because that will only send the one top-level cpp file to the remote host, so it will be missing all the dependencies on the remote machine.

**Comment by poppolopoppo** [comment]
I tried to remove /ZW detection, without surprise this is not working.

Changed ObjectNode.cpp to don't strip /ZW from preprocessor pass, and added vcmeta.dll to compiler extra files. The build is correctly executed on remote workers, but compilation fails on #using (/FU) references and would probably fail after that on /AI which specifies where to look for CLR metadata.

I think the current design of FBuild can't really handle that for now, as it would require some sort on SDK declaration (a 'SDK()' function like  'Compiler()' maybe) due to /ZW mixing compilation and a sort of pre-linking pass for resolving #using directives...

So I guess we'll look for limiting /ZW usage as you suggested, but we'll most likely hit another wall as Unreal isn't really bothering itself with that.

Anyway, thank you for your feedback 😉
